sumpy/recurrence.py

```diff
@@ -0,0 +1,607 @@
+    """
+    final_recurrence = 0
+    # Outer loop is derivative direction
+    # Inner is polynomial order of x_0
+    for m, _ in enumerate(coeffs):
+        for p, _ in enumerate(coeffs[m]):
+            final_recurrence += coeffs[m][p] * _auto_product_rule_single_term(
+                p, m, var)
+    return final_recurrence
+
+
+def recurrence_from_pde(pde: LinearPDESystemOperator) -> sp.Expr:
+    r"""
+    A function that takes in as input a sympy PDE and outputs a recurrence
+    relation.
+
+    :arg pde: a :class:`sumpy.expansion.diff_op.LinearSystemPDEOperator`
+        that must satisfy ``pde.eqs == 1`` and have polynomial coefficients
+        in the coordinates.
+    :arg var: array of sympy variables :math:`[x_0, x_1, \dots]`
+    """
+    ode_in_r, var, ode_order = pde_to_ode_in_r(pde)
+    ode_in_x = ode_in_r_to_x(ode_in_r, var, ode_order).simplify()
+    ode_in_x_cleared = (ode_in_x * var[0]**(ode_order+1)).simplify()
+    # ode_in_x_cleared shouldn't have rational function coefficients
+    assert sp.together(ode_in_x_cleared) == ode_in_x_cleared
+    f_x_derivs = _make_sympy_vec("f_x", ode_order+1)
+    poly = sp.Poly(ode_in_x_cleared, *f_x_derivs)
+    coeffs = ode_in_x_to_coeff_array(poly, ode_order, var)
+    return recurrence_from_coeff_array(coeffs, var)
+
+
+def process_recurrence_relation(r: sp.Expr,
+                                replace=True) -> tuple[int, sp.Expr]:
+    r"""
+    A function that takes in as input a recurrence and outputs a recurrence
+    relation that has the nth term in terms of the n-1th, n-2th etc.
+    Also returns the order of the recurrence relation.
+
+    If replace=True then the recurrence is output in a form that is ideal
+    for pymbolic processing. If replace=False then a standard recurrence 
+    is output.
+
+    :arg recurrence: a recurrence relation in :math:`s(n)`
+    """
+    terms = list(r.atoms(sp.Function))
+    terms = np.array(terms)
+
+    # Sort terms and create idx_l
+    idx_l = []
+    for i in range(len(terms)):
+        tms = list(terms[i].atoms(sp.Number))
+        if len(tms) == 1:
+            idx_l.append(tms[0])
+        else:
+            idx_l.append(0)
+    idx_l = np.array(idx_l, dtype='int')
+    idx_sort = idx_l.argsort()
+    idx_l = idx_l[idx_sort]
+    terms = terms[idx_sort]
+
+    # Order is the max difference between highest/lowest in idx_l
+    order = max(idx_l) - min(idx_l) + 1
+
+    # How much do we need to shift the recurrence relation
+    shift_idx = max(idx_l)
+
+    # Get the respective coefficients in the recurrence relation from r
+    n = sp.symbols("n")
+    s = sp.Function("s")
+    coeffs = sp.poly(r, list(terms)).coeffs()
+
+    # Re-arrange the recurrence relation so we get s(n) = ____
+    # in terms of s(n-1), ...
+    true_recurrence = sum([coeffs[i]/coeffs[-1] * terms[i]
+                           for i in range(0, len(terms)-1)])
+    true_recurrence1 = true_recurrence.subs(n, n-shift_idx)
```

Comment:
Why not solve for the highest-order "s"?

Reply:
Good suggestion, but after looking at this I would not want the denominators to collect after calling solve since I need to perform a re-index later. I want to keep the s(i) terms separate.